calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -56,78 +83,213 @@ type ResourceActionCommand func(*client.Client, unversioned.Resource) (unversion
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            false,
-		apply:             apply,
+		isNamespaced:      isNamespaced,
 		create:            create,
 		update:            update,
 		delete:            delete,
+		get:               get,
 		list:              list,
 	}
-	helpers[tmd] = rh
+	helpers[res.GetObjectKind().GroupVersionKind()] = rh
 
-	tmd = resList.GetTypeMetadata()
 	rh = resourceHelper{
-		typeMetadata:      tmd,
+		listResource:      resList.(ResourceListObject),
 		resourceType:      reflect.ValueOf(resList).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            true,
 	}
-	helpers[tmd] = rh
+	helpers[resList.GetObjectKind().GroupVersionKind()] = rh
+
+	for _, v := range names {
+		kindToRes[v] = res
+	}
+}
+
+func (rh resourceHelper) GetObjectType() reflect.Type {
+	return rh.resourceType
+}
+
+// Apply is an un-typed method to apply (create or update) a resource. This calls Create
+// and if the resource already exists then we call the Update method.
+func (rh resourceHelper) Apply(client client.Interface, resource runtime.Object) (runtime.Object, error) {
+
+	// Try to assert the resource type and remove the resourceVersion, because
+	// Create call can't have resourceVersion set and Update automatically
+	// gets and sets the resourceVersion to the latest one from the datastore.
+	_, ok := resource.(ResourceObject)
+	if ok {
+		resource.(ResourceObject).GetObjectMeta().SetResourceVersion("")
+	} else {
+		resource.(ResourceListObject).GetListMeta().SetResourceVersion("")
+	}
+
+	// Try to create the resource first.
+	ro, err := rh.Create(client, resource)
+
+	switch err.(type) {
+	case cerrors.ErrorResourceAlreadyExists:
+		// Try updating if the resource already exists.
+		return rh.Update(client, resource)
+	default:
+		// For any other errors, return the error
+		return ro, err
+	}
+}
+
+// Create is an un-typed method to create a new resource.  This calls directly
+// through to the resource helper specific Create method which will map the untyped call to
+// the typed interface on the client.
+func (rh resourceHelper) Create(client client.Interface, resource runtime.Object) (runtime.Object, error) {
+	return rh.create(client, resource.(ResourceObject))
+}
+
+// Update is an un-typed method to update an existing resource. This calls the resource
+// specific Get method to get the resourceVersion, and then calls resource specific
+// Update method with the resource with the updated resourceVersion.
+// We retry 5 times if there is an update conflict during the Update operation.
+func (rh resourceHelper) Update(client client.Interface, resource runtime.Object) (runtime.Object, error) {
+	var err error
+	for i := 0; i < 5; i++ {
+
+		// Get the resource to get the resourceVersion.
+		ro, err := rh.get(client, resource.(ResourceObject))
```

Comment:
Shouldn't we only get the resource version if it wasn't explicitly supplied in the resource?

I assume if the YAML contains the ResourceVersion and we do an update, we expect calicoctl to honor that?

Reply:
Which means you probably need to keep track of whether the resource version was initially supplied so that you don't keep retrying in that particular case.

Reply:
not sure I agree. K8s, for example, just ignores ResourceVersion if it's provided for Apply/Update. What is the use case for honoring it for update? I feel like it'll just cause a lot of user issues when users try to save the yaml output and then use apply to update their resources and we reject it because the ResourceVersion is not the latest.

Reply:
Actually, I take that back. I just tested it, they ignore it for Create it seems like, but not for Update/Apply